elixir/lib/contracts.ex

```diff
@@ -69,7 +157,63 @@ defmodule Trento.Contracts do
     end
   end
 
-  defp decode(type, data) do
+  defp verify_event_signature(
+         %CloudEvents.CloudEvent{
+           data: {:proto_data, %Google.Protobuf.Any{value: data}},
+           attributes: %{
+             "time" => %CloudEvents.CloudEventAttributeValue{
+               attr: {:ce_timestamp, %{seconds: time}}
+             },
+             "expire_at" => %CloudEvents.CloudEventAttributeValue{
+               attr: {:ce_timestamp, %{seconds: expire_at}}
+             },
+             "signature" => %CloudEvents.CloudEventAttributeValue{attr: {:ce_bytes, signature}}
+           }
+         } = event,
+         public_key
+       ) do
+    signing_key =
+      public_key
+      |> :public_key.pem_decode()
+      |> Enum.at(0)
+      |> :public_key.pem_entry_decode()
+
+    event_signature_valid? =
+      :public_key.verify("#{data}#{time}#{expire_at}", :sha256, signature, signing_key)
+
+    if event_signature_valid? do
+      {:ok, event}
+    else
+      {:error, :invalid_event_signature}
+    end
+  end
+
+  defp verify_event_signature(_, _), do: {:error, :invalid_event_signature}
+
+  defp verify_event_validity(
+         %CloudEvents.CloudEvent{
+           attributes: %{
+             "expire_at" => %CloudEvents.CloudEventAttributeValue{
+               attr: {:ce_timestamp, %{seconds: expire_at_ts}}
+             }
+           }
+         } = event
+       ) do
+    expire_at = DateTime.from_unix!(expire_at_ts)
+    event_valid? = DateTime.compare(DateTime.utc_now(), expire_at) == :lt
+
+    if event_valid? do
+      {:ok, event}
+    else
+      {:error, :event_expired}
+    end
+  end
+
+  defp verify_event_validity(_), do: {:error, :event_expired}
```

Comment:
nitpick: maybe we should return a different error `expire_at_not_received`
not a big deal, but just to maybe differenciate between errors.
I'm not even sure of this though